flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/TableSinkITCase.scala

```diff
@@ -635,6 +635,58 @@ class TableSinkITCase extends StreamingTestBase {
     assertEquals(expected.sorted, result.sorted)
   }
 
+  @Test
+  def testCharPrecisionEnforcer(): Unit = {
+    val dataId = TestValuesTableFactory.registerData(data3)
+    tEnv.executeSql(
+      s"""
+         |CREATE TABLE char_src (
+         |  id INT,
+         |  shopId BIGINT,
+         |  txt CHAR(9)
+         |) WITH (
+         |  'connector' = 'values',
+         |  'data-id' = '$dataId'
+         |)
+         |""".stripMargin)
+    tEnv.executeSql(
+      s"""
+         |CREATE TABLE char_sink (
+         |  id INT,
+         |  shopId BIGINT,
+         |  txt CHAR(9)
+         |) WITH (
+         |  'connector' = 'values',
+         |  'sink-insert-only' = 'true'
+         |)
+         |""".stripMargin)
+
+    // default should fail, because there are char values exceeding the length specified
+    try {
+      tEnv.executeSql("INSERT INTO char_sink SELECT * FROM char_src").await()
+      fail("Execution should fail.")
+    } catch {
+      case t: Throwable =>
+        val exception = ExceptionUtils.findThrowableWithMessage(
+          t,
+          "Column 'txt' is of CHAR/VARCHAR type with a length of '9', however, a string " +
```

Comment:
Maybe this was a misunderstanding. But I thought that the discussion is only about trimming, not about throwing errors. Because we as Flink can currently not guarantee perfect trimming, so this exception might not necessarily be the fault of the user.

Reply:
True, changed to just to values `IGNORE` and `TRIM`